### ISSUE TITLE
Add response validation for non 200 responses when using Swagger 2.0

### ISSUE THREAD

**Issue by striglia** [striglia/pyramid_swagger#143]
tween.py currently contains this logic:

```python
    # Don't attempt to validate non-success responses in 1.2
    if not 200 <= response.status_code <= 203:
        return
```

This obviously applies to 2.0 as well, despite 2.0 having much improved support for specifying the content of these responses. We should absolutely update the comment, and at least consider also having this logic check for the swagger version and validate error responses if it's >= 2.0.

**Comment by prat0318** [comment]
> This obviously applies to 2.0 as well

No, it does not. There are separate response handlers for [1.2](https://github.com/striglia/pyramid_swagger/blob/master/pyramid_swagger/tween.py#L473) and [2.0](https://github.com/striglia/pyramid_swagger/blob/master/pyramid_swagger/tween.py#L526) version (the name of the methods though easily can create the confusion). The snippet quoted is from `1.2` handler which stands correct.

And earlier i did check that `bravado-core` correctly validated the response bodies against the swagger-spec for all responses (not sure why, but on checking now, an exception in application like `raise pyramid.httpexceptions.HTTPNotFound()` was raising an exception bypassing pyramid-swagger, should be something wrong in my setup).

A subtle point to note though is that request and response validation error responses (400 and 500 resp.) raised by `bravado-core` are skipped from checking its format again (by bravado-core). This might lead to bugs where
application developer writes a swagger spec with error response matching format eventually raised by bravado-core. But actually that format is ignored and the one actually checked (application level error format) remains unmentioned. I am not sure i am clear enough on this (the scenario is a bit complex to describe).

**Comment by striglia** [comment]
Ah! Then is the value of this ticket just "refactor function names to be clearer about which version they validate for"?

I'm definitely not following the subtle point you raise. Maybe walk through an example? Can also be another ticket.

**Comment by prat0318** [comment]
I will try with an example.

Suppose there is an endpoint like so:
```
    "/hours/{biz_id}": {
      "get": {
        "responses": {
          "200": {
            "description": "Week hours",
            "schema": {
              "$ref": "#/definitions/WeekHours"
            }
          },
          "400": {
            "description": "Invalid input"
          }
        },
        "summary": "Get happy hour times by biz_id",
        "operationId": "getHours",
        "tags": ["hours"],
        "parameters": [
          {
            "in": "path",
            "description": "Business ID",
            "name": "biz_id",
            "required": true,
            "type": "integer"
          }
        ]
      },
```

Now, when i test the endpoint for errors and call `/hours/some_string`, i get a 400 raised by `bravado_core` with message of type `string` : 
```
 u'some_string' is not of type 'integer' Failed validating 'type' in schema: {'description': 'Business ID', 'in': 'path', 'name': 'biz_id', 'required': True, 'type': 'integer'} On instance: u'some_string' 
```

I happily change my error response format to : 
```
          "400": {
             "description": "Invalid input",
+           "schema": {"type": "string"}
          }
```

Later in the code, when i add some application level request check, say, `id` should only be odd. i raise a `raise HttpClientError({'reason': 'id must be odd.'})` where the message type is now `dict`/`json`.

Now, `bravado-core` will complain for `/hours/4` that the response type for 400 must be `string` instead of `object`. Though the response raised as `string` was never validated against the type (because it was not application level error but `bravado-core` level error). So, there are two types of error types being raised here (and the first one is kind of not known to the application developer, and hence can lead to subtle errors).

**Comment by wting** [comment]
Ran into this issue this past week.

1. Endpoint raises 400.
2. 400 is not documented in Swagger for that endpoint.
3. `pyramid_swagger` happily returns the response.
4. Bravado client dies with the message:

> *** HTTPError: 400 Bad Request: Response specification matching http status_code 400 not found for operation Operation(my_operation_name). Either add a response specification for the status_code or use a `default` response.

**Comment by striglia** [comment]
Ah. Sounds like we should:

1. Add a test that omits a non200 status in the response and asserts we raise an error and
2. Fix the behavior.

I think the simplest, most consistent behavior is to return a 500 in these cases (even if you were attempting to return a 4xx status code). It is an issue server-side (the spec isn't right!) and looks properly alarming to clients.
